Respect $XDG_DATA_HOME if set

ccjmne opened the pull request:
Respect the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), which was adopted by Timewarrior in version 1.5.0, according to [the release notes](https://github.com/gothenburgbitfactory/timewarrior/blob/1072700d5811ef4105636cdea40f6045b492eac8/ChangeLog#L115).

This commit make the completion system look for the data directory in these three places (in that order):

    1. $XDG_DATA_HOME/timewarrior/data/
    2. $HOME/.local/share/timewarrior/data/
    3. $HOME/.timewarrior/data/

Directories 2 and 3 were always looked up, whereas the first one is new and prioritised.

ccjmne committed:
Respect $XDG_DATA_HOME if set

Respect the XDG Base Directory Specification [1], which was adopted by
Timewarrior in version 1.5.0, according to the release notes [2].

This commit make the completion system look for the data directory in
the priority order used by Timewarrior itself, as of 1.8.0:

    1. $HOME/.timewarrior/data/
    2. $XDG_DATA_HOME/timewarrior/data/
    3. $HOME/.local/share/timewarrior/data/

[1]: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
[2]: https://github.com/gothenburgbitfactory/timewarrior/blob/1072700d5811ef4105636cdea40f6045b492eac8/ChangeLog#L115

Reviewed-by: Ian Kenney <ianmichaelkenney@gmail.com>

ccjmne commented:
Would close #5.

ccjmne commented:
We may want to drop the second commit, which only tries to restore the original lookup order for possible data directory locations.

I originally presumed that the one introduced in 657ce6c is actually preferrable, but it may throw off people that have an existing yet empty directory at `$HOME/.local/share/timewarrior/data` as well as an active one at `$HOME/.timewarrior/data`.

Just let me know, I'll update accordingly :)

ianmkenney commented:
Thanks for the PR @ccjmne! I generally like the solution you implemented in 657ce6c. From what I can see, even if `$HOME/.local/share/timewarrior/data` exists and is empty, and `~/.timewarrior` exists, the later is used. It even goes as far as generating the `~/.timewarrior/data/` directory if it wasn't already there. Even setting `XDG_DATA_HOME` when this directory exists doesn't override the behavior. See the shell output below from a fresh install of 1.7.1, which should have the same behavior WRT data as 1.8.0.

```
root@1300433972ed:~# ls .local/share/timewarrior
ls: cannot access '.local/share/timewarrior': No such file or directory
root@1300433972ed:~# ls .timewarrior                 
ls: cannot access '.timewarrior': No such file or directory
root@1300433972ed:~# timew    
Create new database in /root/.local/share/timewarrior? (yes/no) yes

Welcome to Timewarrior.

There is built-in help:
    timew help
    timew help <command>
    (and more)

There is a fully-detailed man page:
    man timew

root@1300433972ed:~# ls .local/share/timewarrior/data/
tags.data
root@1300433972ed:~# timew start
Tracking 
  Started 2025-07-31T12:59:20
  Current                  20
  Total               0:00:00
root@1300433972ed:~# timew stop
Recorded 
  Started 2025-07-31T12:59:20
  Ended                    29
  Total               0:00:09
root@1300433972ed:~# timew summary

Wk  Date       Day Tags    Start      End    Time   Total
W31 2025-07-31 Thu      12:59:20 12:59:29 0:00:09 0:00:09
                                                         
                                                  0:00:09

root@1300433972ed:~# mkdir .timewarrior
root@1300433972ed:~# timew summary
No filtered data found in the range 2025-07-31T00:00:00 - 2025-08-01T00:00:00.
root@1300433972ed:~# XDG_DATA_HOME=./tmp/xdg_data/ timew summary
No filtered data found in the range 2025-07-31T00:00:00 - 2025-08-01T00:00:00.
root@1300433972ed:~# rm -r .timewarrior/
root@1300433972ed:~# XDG_DATA_HOME=./tmp/xdg_data/ timew summary
Create new database in ./tmp/xdg_data//timewarrior? (yes/no) yes
No filtered data found in the range 2025-07-31T00:00:00 - 2025-08-01T00:00:00.
```

So I think your solution in 657ce6c would work great with the only change being that the check for the legacy path is in a separate `if` block. If you have any other thoughts please let me know!

ccjmne commented:
Excellent catch and beautiful proof, @ianmkenney!

I replicated it, 'cause the exercise seems like a nice habit I should develop, and could observe exactly what you described with `1.8.0` as well.

So the lookup order goes (in descending order of priority):

- `~/.timewarrior`
- `$XDG_DATA_HOME/timewarrior`
- `~/.local/data/timewarrior`

I'll adjust it accordingly right away.

ccjmne commented:
Should be good to go.
I also surrounded the paths in the `[ -d "$HOME/something-something" ]` tests, as you never know who wants a whitespace in the name of their homedir, and you'd already quoted these paths in the subsequent assignments.

Just to make sure, I still did the double-check, and it turns out I'm completely out of the loop and **this** is the reality:

```sh
asdf='as df'; mkdir $asdf; [ -d $asdf ] && echo found
found
```

Live and learn... That's only in `zsh`, however: `bash` and `sh` both fail:
```
sh-5.3$ asdf='as df'; [ -d $asdf ] && echo found
sh: [: as: binary operator expected
```

But hey, it's a zsh-completion script! So I force-pushed once more, undoing any superflous changes, and the diff is now tidy: just the simple, obvious change.
Thanks again :)

ianmkenney commented:
This looks great. Thanks for taking the time to make the completion script consistent with the application behavior! I'll push these upstream in the next couple of days.

ccjmne commented:
Thank you!
If you'd like, I already prepared last night [a commit](https://github.com/ccjmne/timewarrior/commit/35f70f631c498466a8cf952c6d6f380f65100311) that encapsulates the changes since the last time this completion module was merged to Timewarrior itself, and can open a PR anytime.

If you'd rather do it yourself though, it's just as fine by me!

ccjmne commented:
Oh, I'm only catching up now, and I see they're suggesting you don't even look at the data directory... alright alright then. Looking forward to it :)

ianmkenney commented:
Going to wait and see what they say and then from there remove the data directory dependency. If we can go ahead and add `dom.tracked.n.tags`, I'd rather just wait until that's in place.

Again, many thanks for checking into and fixing the `XDG_DATA_HOME` variable issue. The conversation over at `timewarrior` wouldn't have happened otherwise!
